src/main/java/dev/jbang/util/ConsoleInput.java

```diff
@@ -1,41 +1,78 @@
 package dev.jbang.util;
 
-import java.util.concurrent.ExecutionException;
-import java.util.concurrent.ExecutorService;
-import java.util.concurrent.Executors;
-import java.util.concurrent.Future;
-import java.util.concurrent.TimeUnit;
-import java.util.concurrent.TimeoutException;
-
-public class ConsoleInput {
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.util.concurrent.*;
+
+public abstract class ConsoleInput {
 	private final int tries;
 	private final int timeout;
 	private final TimeUnit unit;
 
-	public ConsoleInput(int tries, int timeout, TimeUnit unit) {
+	/**
+	 * Will either return a ConsoleInput that enables reading a line from the
+	 * console (using stdin or a tty) or it will return <code>null</code> if no
+	 * console is available.
+	 */
+	public static ConsoleInput get(int tries, int timeout, TimeUnit unit) {
+		if (Util.haveConsole()) {
+			return stdin(tries, timeout, unit);
+		} else if (!Util.isWindows()) {
+			return tty(tries, timeout, unit);
+		} else {
+			return null;
+		}
+	}
+
+	/**
+	 * Returns a regular ConsoleInput based upon System.in
+	 */
+	public static ConsoleInput stdin(int tries, int timeout, TimeUnit unit) {
+		return new ConsoleInput(tries, timeout, unit) {
+			@Override
+			protected Callable<String> readerTask() {
+				return new ConsoleInputReadTask(System.in);
+			}
+		};
+	}
+
+	/**
+	 * Returns a ConsoleInput based upon /dev/tty which only works on Linux and Mac.
+	 */
+	public static ConsoleInput tty(int tries, int timeout, TimeUnit unit) {
+		return new ConsoleInput(tries, timeout, unit) {
+			@Override
+			protected Callable<String> readerTask() throws IOException {
+				return new ConsoleInputReadTask(new FileInputStream("/dev/tty"));
```

Comment:
there is not a case where /dev/tty can be unreadable and we need to still use UI?